include/aws/common/task_scheduler.h

```diff
@@ -17,48 +17,55 @@
  */
 
 #include <aws/common/common.h>
+#include <aws/common/linked_list.h>
 #include <aws/common/priority_queue.h>
 
+struct aws_task;
+
 typedef enum aws_task_status {
     AWS_TASK_STATUS_RUN_READY,
     AWS_TASK_STATUS_CANCELED,
 } aws_task_status;
 
 /**
- * Pointer to the scheduled function
+ * A scheduled function.
  */
-typedef void(aws_task_fn)(void *arg, enum aws_task_status);
+typedef void(aws_task_fn)(struct aws_task *task, void *arg, enum aws_task_status);
 
-/**
- * Monotonic clock function
+/*
+ * A task object.
+ * Once added to the scheduler, a task must remain in memory until its function is executed.
  */
-typedef int(aws_task_scheduler_clock_fn)(uint64_t *);
-
 struct aws_task {
     aws_task_fn *fn;
     void *arg;
+    uint64_t timestamp;
+    struct aws_linked_list_node node;
 };
 
+static inline void aws_task_init(struct aws_task *task, aws_task_fn *fn, void *arg) {
+    task->fn = fn;
+    task->arg = arg;
+    task->timestamp = 0;
+    aws_linked_list_node_reset(&task->node);
+}
+
 struct aws_task_scheduler {
     struct aws_allocator *alloc;
-    struct aws_priority_queue queue;
-    aws_task_scheduler_clock_fn *clock;
-    uint64_t min_run_time;
+    struct aws_priority_queue timed_queue; /* Tasks scheduled to run at specific times */
+    struct aws_linked_list asap_list;      /* Tasks scheduled to run as soon as possible */
+    struct aws_linked_list running_list;   /* Tasks currently being executed */
```

Comment:
This can be local to the `s_run_all` function.

Reply:
deeeerrrrrp me